testclock/clock_test.go

```diff
@@ -170,6 +170,94 @@ func (*clockSuite) TestNewTimerReset(c *gc.C) {
 	}
 }
 
+func (*clockSuite) TestNewTimerAsyncReset(c *gc.C) {
+	t0 := time.Now()
+	clock := testclock.NewClock(t0)
+	timer := clock.NewTimer(time.Hour)
+	stop := make(chan struct{})
+	var wg sync.WaitGroup
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		select {
+		case <-stop:
+			return
+		case t := <-timer.Chan():
+			c.Errorf("timer accidentally ticked at: %v", t)
+		case <-time.After(testing.LongWait):
+			c.Errorf("test took too long")
+		}
+	}()
+	// Just our goroutine, but we don't go so far as to trigger the wakeup.
+	clock.WaitAdvance(1*time.Minute, 10*time.Millisecond, 1)
+	// Reset shouldn't trigger a wakeup, just move when it thinks it will wake up.
+	timer.Reset(time.Hour)
+	clock.WaitAdvance(1*time.Minute, 10*time.Millisecond, 1)
+	timer.Reset(time.Minute)
+	clock.WaitAdvance(30*time.Second, 10*time.Millisecond, 1)
+	// Now tell the goroutine to stop and start another one that *does* want to
+	// wake up
+	close(stop)
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		select {
+		case t := <-timer.Chan():
+			c.Logf("timer successfully ticked: %v", t)
+		case <-time.After(1 * time.Second):
+			c.Errorf("timer took too long")
+		}
+	}()
+	// And advance the clock long enough to cause it to notice
+	clock.WaitAdvance(30*time.Second, 10*time.Millisecond, 1)
+	wg.Wait()
+}
+
+func (*clockSuite) TestNewTimerResetCausesWakeup(c *gc.C) {
+	t0 := time.Now()
+	clock := testclock.NewClock(t0)
+	timer1 := clock.NewTimer(time.Hour)
+	timer2 := clock.NewTimer(time.Hour)
+	timer3 := clock.NewTimer(time.Hour)
+	var wg sync.WaitGroup
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		select {
+		case t := <-timer1.Chan():
+			c.Check(t0, gc.Equals, t)
+		case <-time.After(1 * time.Second):
```

Comment:
one second is perhaps to short for slow architectures and noisy neighbours? Maybe?

Reply:
I found that we had constants 'longWait' and 'shortWait' in the file already, so I just switched to those.
If we have to change them we can, but they should be fine. (The test runs in <1ms here, so a noisy machine would have to be 1000x slower).